packages/backend/src/services/PromoteService/PromoteService.ts

```diff
@@ -0,0 +1,761 @@
+        super();
+        this.lightdashConfig = args.lightdashConfig;
+        this.analytics = args.analytics;
+        this.savedChartModel = args.savedChartModel;
+        this.projectModel = args.projectModel;
+        this.spaceModel = args.spaceModel;
+        this.dashboardModel = args.dashboardModel;
+    }
+
+    private async trackAnalytics(
+        user: SessionUser,
+        event: 'promote.executed' | 'promote.error',
+        promotedContent: PromotedChart | PromotedDashboard,
+        upstreamContent: UpstreamChart | UpstreamDashboard,
+        error?: string,
+    ) {
+        this.analytics.track({
+            event,
+            userId: user.userUuid,
+            properties: {
+                chartId:
+                    'chart' in promotedContent
+                        ? promotedContent.chart.uuid
+                        : undefined,
+                dashboardId:
+                    'dashboard' in promotedContent
+                        ? promotedContent.dashboard.uuid
+                        : undefined,
+                fromProjectId: promotedContent.projectUuid,
+                toProjectId: upstreamContent.projectUuid,
+                organizationId: user.organizationUuid!,
+                slug:
+                    'chart' in promotedContent
+                        ? promotedContent.chart.slug
+                        : promotedContent.dashboard.slug,
+                withNewSpace: upstreamContent.space === undefined,
+                hasExistingContent:
+                    'chart' in upstreamContent
+                        ? upstreamContent.chart !== undefined
+                        : upstreamContent.dashboard !== undefined,
+                chartsCount:
+                    'dashboard' in promotedContent
+                        ? promotedContent.chartUuids.length
+                        : undefined,
+                error,
+            },
+        });
+    }
+
+    private async getPromoteCharts(
+        user: SessionUser,
+        upstreamProjectUuid: string,
+        chartUuid: string,
+    ): Promise<{
+        promotedChart: PromotedChart;
+        upstreamChart: UpstreamChart;
+    }> {
+        const savedChart = await this.savedChartModel.get(chartUuid, undefined);
+
+        const promotedSpace = await this.spaceModel.getSpaceSummary(
+            savedChart.spaceUuid,
+        );
+        const upstreamCharts = await this.savedChartModel.find({
+            projectUuid: upstreamProjectUuid,
+            slug: savedChart.slug,
+        });
+        if (upstreamCharts.length > 1) {
+            throw new AlreadyExistsError(
+                `There are multiple charts with the same identifier ${savedChart.slug}`,
+            );
+        }
+        const upstreamChart =
+            upstreamCharts.length === 1 ? upstreamCharts[0] : undefined;
+        const upstreamSpaces = await this.spaceModel.find({
+            projectUuid: upstreamProjectUuid,
+            slug: promotedSpace.slug,
```

Comment:
just to clarify: we search for the `space` slug of the `promoted` chart, not the `space` of the `existing` chart in the upstream project

This means, if the chart was moved in the upstream project to another space, we will overwrite this change, and restore the `promoted` space. I think this is what we want.